### PR TITLE
De-share the Handler struct in core API

### PR DESCRIPTION
**Pull request by thockin**
An upcoming PR (#102162) adds a handler that only applies on one of these paths.
Having fields that don't work seems bad.

This never should have been shared.  Lifecycle hooks are like a "write" while probes are more like a "read". HTTPGet and TCPSocket don't really make sense as lifecycle hooks (but I can't take that back). When we add gRPC, it is EXPLICITLY a health check (defined by gRPC) not an arbitrary RPC - so a probe makes sense but a hook does not.
    
In the future I can also see adding lifecycle hooks that don't make sense as probes.  E.g. 'sleep' is a common lifecycle request. The only option is `exec`, which requires having a sleep binary in your image.

/kind cleanup

```release-note
NONE
```

cc @yuzhiquan

### CODE REVIEW

**Comment by aojea** [comment]
+1

**Comment by yuzhiquan** [comment]
Bravo

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -2188,9 +2188,29 @@ type Container struct {
 	TTY bool
 }
 
-// Handler defines a specific action that should be taken
+// ProbeHandler defines a specific action that should be taken in a probe.
+// This type has a strong relationship to LifecycleHandler - overlapping fields
```
Comment (SergeyKanzhelev) [inline pkg/apis/core/types.go:2192]: why the comment? Is it a requirement to keep them identical in the long run? Even implementation is different (see https://github.com/kubernetes/kubernetes/pull/86139/).

Comment (SergeyKanzhelev) [reply pkg/apis/core/types.go:2192]: followed up with https://github.com/kubernetes/kubernetes/pull/106285

**Pull request merged**